src/appleseed/foundation/utility/poison.h

```diff
@@ -82,6 +85,12 @@ class PoisonImpl
     }
 };
 
+template <typename T>
+inline void always_poison(T& x)
+{
+    PoisonImpl<T>::do_always_poison(x);
```

Comment:
You should simply use `PoisonImpl<T>::do_poison(x)` here. That way we don't have to add duplicates of all the poisoning methods.

Reply:
I didn't do that on purpose. You never know how encapsuled the templates are. Let's say I'm using a `Vector<Vector<float, 3>>` for whatever reason (it's not the best example). In that case, the outer Vector would be called to do the `always_poison` and the inner Vector would still be stuck in the regular `poison`, resulting in no poison.
I'm relatively unhappy with this doubled infrastructure as well, but I don't want to restrict users to use built-in types at the first template-level.

Reply:
I'm not sure to understand your argument.

`poison<T>()` is just a wrapper for `PoisonImpl<T>::do_poison(x)`, except that it doesn't do anything in debug. All we want is a new variant of it that _always_ call `PoisonImpl<T>::do_poison(x)`, regardless of the configuration.

The poisoning itself is done by `PoisonImpl<T>` and it must do the same whether it's called via `debug_poison()` or `always_poison()`.

Reply:
Let's unwrap the chain:
In `src/appleseed/renderer/modeling/bssrdf/randomawlkbssrdf.cpp:242` the function `always_poison` gets called. This results in the following chain
```
foundation/utility/poison.h:89 (always_poison)
foundation/math/vector.h:507 (PoisonImpl<Vector<float, 3>>::do_always_poison)
foundation/utility/poision.h:125 (PoisonImpl<float>::do_always_poison)
```
In this scenario it doesn't make a difference if we call `do_always_poison` or `do_poison`, because for floats there is no debug macro (floats will always be poisoned, regardless of the build type).
Let's assume I got a `Vector<Vector<float, 3>>` and call `poison` on it. Then the call chain would look like this:
```
foundation/utility/poison.h:89 (always_poison)
foundation/math/vector.h:507 (PoisonImpl<Vector<Vector<float, 3>, 3>>::do_always_poison)
foundation/utility/poison.h:89 (always_poison)
foundation/math/vector.h:507 (PoisonImpl<Vector<float, 3>>::do_always_poison)
foundation/utility/poison.h:89 (always_poison)
foundation/utility/poision.h:125 (PoisonImpl<float>::do_always_poison)
```
This results in all floats of the inner vector to be poisoned.

But if we exchange the `do_always_poison` call for a regular `do_poison` call (as suguested), the chain looks like this:
```
foundation/utility/poison.h:89 (always_poison)
foundation/math/vector.h:500 (PoisonImpl<Vector<T, N>>::do_poison)
foundation/utility/poison.h:95 (poison)
```
The calls stop there, because the function `poison.h:95 (poison)` has the debug macro in it. We won't run into this problem, as long as the template parameter of Vectors will always be doubles or floats. But as soon as we have more complex structures, leading to a call-chain like the above mentioned one, `always_poison` will not result in a poisoned variable, which I want to prevent. From my point of view an `always` function should do its job **always**. In the above mentioned scenario it wouldn't do so.

I highly agree that this kind of double-structure is nasty and I'm thinking about how to get rid of it, without running into trouble.

The codelines of this comment refer to the as-is state in this branch. I haven't changed the `poison` functions to `debug_poison` yet. Also, I left out the `src/appleseed/` in the mutli-line structure, to increase readability.

Reply:
Oh, good point, I understand now.

Maybe a nicer solution would be to ask that all `PoisonImpl` specializations use `always_poison()` in their implementation?

Reply:
Done (Details below).